bee-network/bee-autopeering/src/peer/mod.rs

```diff
@@ -0,0 +1,334 @@
+    }
+
+    /// Returns the services of this peer.
+    pub fn services(&self) -> &ServiceMap {
+        &self.services
+    }
+
+    /// Sets the services of this peer all at once.
+    pub(crate) fn set_services(&mut self, services: ServiceMap) {
+        self.services = services;
+    }
+
+    /// Returns whether the peer provides a corresponding service.
+    pub fn has_service(&self, service_name: impl AsRef<str>) -> bool {
+        self.services.get(service_name).is_some()
+    }
+
+    /// Adds a service with address binding to this peer.
+    pub fn add_service(&mut self, service_name: impl ToString, protocol: ServiceProtocol, port: u16) {
+        self.services.insert(service_name.to_string(), protocol, port);
+    }
+
+    /// Returns the address, i.e. the `libp2p_core::Multiaddr`, belonging to the given service name.
+    ///
+    /// Example: "peering" => `/ip4/127.0.0.1/udp/14627`.
+    pub fn get_service_multiaddr(&self, service_name: impl AsRef<str>) -> Option<Multiaddr> {
+        if let Some(service) = self.services.get(service_name) {
+            let mut multiaddr = Multiaddr::empty();
+
+            match self.ip_address {
+                IpAddr::V4(ipv4_addr) => multiaddr.push(Protocol::Ip4(ipv4_addr)),
+                IpAddr::V6(ipv6_addr) => multiaddr.push(Protocol::Ip6(ipv6_addr)),
+            };
+
+            multiaddr.push(service.to_libp2p_protocol());
+            Some(multiaddr)
+        } else {
+            None
+        }
+    }
+
+    /// Creates a peer from its Protobuf representation/encoding.
+    pub fn from_protobuf(bytes: &[u8]) -> Result<Self, DecodeError> {
+        Ok(proto::Peer::decode(bytes)?.into())
+    }
+
+    /// Returns the Protobuf representation of this peer.
+    pub fn to_protobuf(&self) -> Result<BytesMut, EncodeError> {
+        let services: proto::ServiceMap = self.services.clone().into();
+
+        let peer = proto::Peer {
+            ip: self.ip_address.to_string(),
+            public_key: self.public_key().as_ref().to_vec(),
+            services: Some(services),
+        };
+
+        let mut buf = BytesMut::with_capacity(peer.encoded_len());
+        peer.encode(&mut buf)?;
+
+        Ok(buf)
+    }
+
+    pub(crate) fn into_id(self) -> PeerId {
+        self.peer_id
+    }
+}
+
+impl fmt::Debug for Peer {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Peer")
+            .field("peer_id", &self.peer_id.to_string())
+            .field("public_key", &bs58::encode(self.public_key().as_ref()).into_string())
+            .field("ip_address", &self.ip_address)
+            .field("services", &self.services.to_string())
+            .finish()
+    }
+}
+
+impl From<proto::Peer> for Peer {
+    fn from(peer: proto::Peer) -> Self {
+        let proto::Peer {
+            public_key,
+            ip,
+            services,
+        } = peer;
+
+        // TODO: resolve DNS addresses
```

Comment:
Status ?

Reply:
I wasn't sure before live-testing whether nodes would exchange DNS addresses, but seems like they only exchange already resolved addresses. So the TODO can be removed.

Done